examples/parser/Parser/Core.roc

```diff
@@ -0,0 +1,292 @@
+alt : Parser input a, Parser input a -> Parser input a
+alt = \first, second ->
+    buildPrimitiveParser \input ->
+        when parsePartial first input is
+            Ok { val: val, input: rest } -> Ok { val: val, input: rest }
+            Err (ParsingFailure firstErr) ->
+                when parsePartial second input is
+                    Ok { val: val, input: rest } -> Ok { val: val, input: rest }
+                    Err (ParsingFailure secondErr) ->
+                        Err (ParsingFailure ("\(firstErr) or \(secondErr)"))
+
+## Runs a parser building a function, then a parser building a value,
+## and finally returns the result of calling the function with the value.
+##
+## This is useful if you are building up a structure that requires more parameters
+## than there are variants of `map`, `map2`, `map3` etc. for.
+##
+## For instance, the following two are the same:
+##
+## >>> const (\x, y, z -> Triple x y z)
+## >>> |> map3 Parser.Str.nat Parser.Str.nat Parser.Str.nat
+##
+## >>> const (\x -> \y -> \z -> Triple x y z)
+## >>> |> apply Parser.Str.nat
+## >>> |> apply Parser.Str.nat
+## >>> |> apply Parser.Str.nat
+##
+## (And indeed, this is how `map`, `map2`, `map3` etc. are implemented under the hood.)
+##
+## # Currying
+## Be aware that when using `apply`, you need to explicitly 'curry' the parameters to the construction function.
+## This means that instead of writing `\x, y, z -> ...`
+## you'll need to write `\x -> \y -> \z -> ...`.
+## This is because the parameters to the function will be applied one-by-one as parsing continues.
+apply : Parser input (a -> b), Parser input a -> Parser input b
+apply = \funParser, valParser ->
+    combined = \input ->
+        { val: funVal, input: rest } <- Result.try (parsePartial funParser input)
+        parsePartial valParser rest
+        |> Result.map \{ val: val, input: rest2 } ->
+            { val: funVal val, input: rest2 }
+
+    buildPrimitiveParser combined
+
+# Internal utility function. Not exposed to users, since usage is discouraged!
+#
+# Runs `firstParser` and (only) if it succeeds,
+# runs the function `buildNextParser` on its result value.
+# This function returns a new parser, which is finally run.
+#
+# `andThen` is usually more flexible than necessary, and less efficient
+# than using `const` with `map` and/or `apply`.
+# Consider using those functions first.
+andThen : Parser input a, (a -> Parser input b) -> Parser input b
+andThen = \firstParser, buildNextParser ->
+    fun = \input ->
+        { val: firstVal, input: rest } <- Result.try (parsePartial firstParser input)
+        nextParser = buildNextParser firstVal
+
+        parsePartial nextParser rest
+
+    buildPrimitiveParser fun
+
+## Try a list of parsers in turn, until one of them succeeds
+oneOf : List (Parser input a) -> Parser input a
+oneOf = \parsers ->
+    List.walkBackwards parsers (fail "oneOf: The list of parsers was empty") (\laterParser, earlierParser -> alt earlierParser laterParser)
```

Comment:
Is there a reason to walk backwards? I think most folks would expect `oneOf` to try parsers in forward list order

Reply:
Oops, disregard. I realized now that `alt` will construct it so that they are tried in forward list order

Reply:
Maybe worth calling out the order parsers will be run in, in the api doc comment of `oneOf`?